### ISSUE TITLE
Support for multi-table atomic transactions

### ISSUE THREAD

**Issue by jhesed** [numberoverzero/bloop#120]
Hi @numberoverzero,

Great project!
Do you plan on supporting multi-table transactions similar to https://aws.amazon.com/blogs/aws/dynamodb-transaction-library/ ?

Something like,

> begin transaction
> put item 1
> put item 2
> ...
> commit transaction

I saw an atomic implementation in _Bloop_, but I'm not sure if it works across multiple tables.

**Comment by numberoverzero** [comment]
---

Getting your note at the end out of the way first, bloop's `atomic=` option for [`Engine.save`](https://bloop.readthedocs.io/en/latest/user/engine.html#save) and [`Engine.delete`](https://bloop.readthedocs.io/en/latest/user/engine.html#delete) provides a **per-item** guarantee, not a guarantee across the collection of items you are saving/deleting.   If your save/delete is successful, the version of the object in DynamoDB was identical to your local state before the operation happened.  This needs to be clarified in the documentation immediately, thanks for pointing it out.

---

# tl;dr

Thanks for the great question!  I got a chance to work with the transaction library while at AWS, and it made a lot of work easier.  It also made some work harder.

I don't think Bloop is the right place for exposing this functionality, but it might be a good base for an additional library, or a bloop-toolbelt, which would help #84.

# justification

There are three primary reasons I'm not comfortable adding this functionality to Bloop.  In no particular order:
1. Bloop tries to preserve visibility into throughput, so you can reason about your Read/Write IOPS without digging through the source code.  A txn requires 7N+4 (N is items in the txn) writes, which easily blows through the free tier even for a txn with one item in each of two tables.
2. Building ACID correctly on top of CD is really hard, and I don't want to get it wrong.  There was a race condition that sat latent for about a year, and then cost our team a couple days to track down.  [Here's the fix](https://github.com/awslabs/dynamodb-transactions/commit/c3470df17469517432133b1f33534795a4657366#diff-a700bce8d2a97c45e27f694a216f5d80).
3. Transactions are sometimes an anti-pattern when using DynamoDB, and forcing them causes performance and operational overhead that can be hard to see until it's too late.

I'm going to break these down, not just for you but because I expect this is a common desire.  I've even looked for a DynamoDB transaction library in python a few times when working on side projects :)

## 1. Preserving throughput visibility

Users should be able to look at a line of code like `engine.save(my_obj)` and know how many calls were made to DynamoDB on their behalf.  Without that information auditing your own performance requirements is basically impossible.  This is already obscured slightly with retries (botocore will retry on certain failures, I don't remember if boto3 will) but we should be able to say `engine.save(*users)` will perform `len(users)` calls and have some confidence.

Once transactions are available a line of code like `txn.commit()` could cause an arbitrary number of writes.  You can still come up with a cost by reviewing all the code before this: count up every update performed in the txn and multiply by ~8.  That's more cognitive load, and I'd rather say "all Bloop operations map to 1 call per object" than "some operations in Bloop are harder to count than others" which forces the user to recall which operations are cheap.

There are a few places where Bloop doesn't hold to this constraint: iterating queries, scans, and streams.  Early on I considered exposing those controls (#58, #91, #107) but they were too ambiguous (LIMIT on the wire is a pre-filter DynamoDB directive but everyone wants a local post-filter directive).  Given that we don't think of iterating a query or stream as a finite unit of work anyway, this seems a fair tradeoff.

A separate library wouldn't be bound by such a design constraint, and could describe performance costs in detail upfront.


## 2. ACID is hard

I'm not confident that I could re-implement dynamodb-tranasctions-library without introducing hard to find bugs, or even re-introducing previously fixed bugs.  As a user I'd feel better about a purpose-built transaction library that wasn't part of a particular ORM, and instead could switch backends eg. migrating from flywheel to pynamodb.  As a developer I'd feel better working on a code base entirely focused on doing transactions correctly; all the utility functions, type annotations, test fixtures can be specifically for making transactions work correctly.  Nothing about streams or backups or weak-ref'd atomic condition building.

The transactions library doesn't support [conditions or return metrics](https://github.com/awslabs/dynamodb-transactions/blob/a3e9f25844cf9097cd9447fbf07e8820cfa12ff4/src/main/java/com/amazonaws/services/dynamodbv2/transactions/Request.java#L174), which are both core pieces of bloop's syntax (I'd argue conditions are **the** reason to prefer bloop, besides maybe streams) and while return metrics can't be enabled (they don't make sense) conditions could be with a lot of work.

I wouldn't want to pioneer that without help.  Maybe in another library :)


## 3. Maybe we don't need ACID

This isn't specific to Bloop, but I don't want Bloop to advance the problem.  Whenever I start reaching for transactions or relationships with DynamoDB (eg. Username and User tables) it's been a sign I should use a relational store.  For small projects I like how fast I can prototype with DynamoDB but the cost of properly doing transactions on a non-relational store immediately exceeds my budget for small projects (<$10/mo in AWS usage).

On larger projects where we've used transactions the operational cost is high; a partial transaction that fails because the host dies leaves cruft that needs to be cleaned up, which means automation to scan the txn table, fix up the state tracking columns for objects.  Another problem is the item size limit of 400kb, which is huge when you've got some guids and booleans but start putting user-driven documents in there and your transactions table starts failing fast.

There are solutions listed in the [design document here](https://github.com/awslabs/dynamodb-transactions/blob/master/DESIGN.md) but to date none have been implemented.  I don't know if that's from lack of necessity or ability, or something else.

**Comment by numberoverzero** [comment]
For what it's worth, if a bloop-specific library did exist to provide transactions, I'd want something like:

* a Transaction class with a context manager that takes an engine so they're loosely coupled eg `with transaction(engine, isolation="committed") as txn:`
* a `ForeignKey` or `Relationship` generic `Type` that takes the model to reference and figures out necessary hash/range key values eg `ref_user = Column(ForeignKey(User))`
* utilities to sweep stuck transactions, break item locks


The second isn't strictly necessary, since relationships could be encoded in the application logic and the library could simply not specify cascading behavior.  It would be enough I think to provide:

```
# models.py
from transactions import make_tx_table
from bloop import BaseModel, Engine

# usual model definition and engine creation
engine = Engine(...)
engine.bind(BaseModel)

# creates a new Model, binds it, and sets engine.txn_model to the new model
make_tx_table(
    engine,
    table_name="txn-integ",
    default_isolation="committed",
    commit_on_close=True,
    abort_on_exc=True
)


# app.py
from transactions import transaction
from models import User, Account, engine

# because commit_on_close is True, this automatically commits when the block exits
# because rollback_on_exc is True, if an exception is raised in the block, the txn is aborted
with transaction(engine) as txn:
    txn.save(user, acct)
```

**Comment by numberoverzero** [comment]
On the specific topic of adding transactions I'm going to close this as won't fix.  I'm happy to hear your ideas for an external implementation, and if you're serious about getting something started I can contribute some code.

On the topic of bloop's `atomic=` keyword for `Engine.save` and `Engine.delete`, I've added and updated some docs in the 2.2 release (#123) which hopefully clarifies things:

* [User Guide - Using the Engine - Save](https://bloop.readthedocs.io/en/latest/user/engine.html#save)

If you think anything's missing or could be improved feel free to comment or re-open.  Thanks for your feedback!

**Comment by numberoverzero** [comment]
@jhesed just a heads up, not sure if you're still using or interested in bloop.

Multi-table transactions are now available in 2.3.0 thanks the the new DynamoDb operations `TransactGetItems` and `TransactWriteItems`:

```
with engine.transaction() as tx:
    tx.save(tweet)
    tx.delete(event)
```
